spring-kafka-test/src/test/java/org/springframework/kafka/test/context/EmbeddedKafkaContextCustomizerTest.java

```diff
@@ -0,0 +1,104 @@
+ */
+public class EmbeddedKafkaContextCustomizerTest {
+
+	/**
+	 * Insures that for each property supplied by {@link EmbeddedKafka#brokerProperties()}, {@link
+	 * org.springframework.core.env.PropertyResolver#resolvePlaceholders(String)} is invoked.
+	 * <p>
+	 * Because {@code EmbeddedKafkaContextCustomizer} encapsulates the creation of {@link
+	 * org.springframework.kafka.test.rule.KafkaEmbedded KafkaEmbedded},
+	 * and because {@code KafkaEmbedded} does not expose much state, it is difficult to directly verify that a property
+	 * string was resolved to a specific value.  Instead, this test insures that a call to {@code
+	 * PropertyResolver#resolvePlaceholders(String)} occurs for each property string returned by {@code EmbeddedKafka
+	 * #brokerProperties()}
+	 * </p>
+	 */
+	@Test
+	public void testBrokerPropertyResolution() {
+
+		// Keys are the properties as they may appear in the @EmbeddedKafka brokerProperties attribute
+		// Values are the result of applying the Spring PropertyResolver.resolvePlaceholders(...) method and are
+		// currently ignored by this test
+		Map<String, String> expectedResults = new HashMap<String, String>() {
+			{
+				put("property.key = ${property.value}", "property.key = resolvedValue");
+				put("foo = bar", "foo = bar");
+			}
+		};
+
+		// Expectations for EmbeddedKafka
+		EmbeddedKafka embeddedKafka = mock(EmbeddedKafka.class);
+		given(embeddedKafka.count()).willReturn(1);
+		given(embeddedKafka.partitions()).willReturn(2);
+		given(embeddedKafka.brokerProperties()).willReturn(
+				new ArrayList<>(expectedResults.keySet()).toArray(new String[]{}));
+
+		// Collaborators and expectations for the application context and environment.
+		DefaultSingletonBeanRegistry beanFactory = mock(DefaultSingletonBeanRegistry.class,
+				withSettings().extraInterfaces(ConfigurableListableBeanFactory.class));
+		ConfigurableApplicationContext context = mock(ConfigurableApplicationContext.class);
+		ConfigurableEnvironment env = mock(ConfigurableEnvironment.class);
+		given(context.getEnvironment()).willReturn(env);
+		given(context.getBeanFactory()).willReturn((ConfigurableListableBeanFactory) beanFactory);
+
+		// Mock the property resolution behavior of PropertyResolver.resolvePlaceholders(...)
+		// by looking up the resolved values from the map
+		given(env.resolvePlaceholders(anyString()))
+				.willAnswer(invocation -> {
+					Object propertyString = invocation.getArgument(0);
+					Assert.isTrue(expectedResults.containsKey(propertyString),
+							"Unexpected property string '" + propertyString + "'"); // sanity
+					return expectedResults.get(propertyString);
+				});
+
+		EmbeddedKafkaContextCustomizer underTest = new EmbeddedKafkaContextCustomizer(embeddedKafka);
+		underTest.customizeContext(context, mock(MergedContextConfiguration.class));
+
+		// Verify that an attempt was made to resolve each property string
+		then(env).should(times(expectedResults.size())).resolvePlaceholders(anyString());
```

Comment:
I somehow don't understand this test - it isn't so simple as you mention.

There is `KafkaStreamsTests` where we can modify its `brokerProperties` like this:
```
brokerProperties = {
				"auto.create.topics.enable=${topics.autoCreate:false}",
				"delete.topic.enable=${topic.delete:true}" }
```
And PP resolver will take care properly about nonexistent properties and fall back to the defaults after column.

Reply:
Hi @artembilan,

When I make the following change:
```diff
diff --git a/spring-kafka/src/test/java/org/springframework/kafka/kstream/KafkaStreamsTests.java b/spring-kafka/src/test/java/org/springframework/kafka/kstream/KafkaStreamsTests.java
index 6014807..4f7ee2a 100644
--- a/spring-kafka/src/test/java/org/springframework/kafka/kstream/KafkaStreamsTests.java
+++ b/spring-kafka/src/test/java/org/springframework/kafka/kstream/KafkaStreamsTests.java
@@ -78,8 +78,8 @@ import org.springframework.util.concurrent.SettableListenableFuture;
                                KafkaStreamsTests.STREAMING_TOPIC2,
                                KafkaStreamsTests.FOOS },
                brokerProperties = {
-                               "auto.create.topics.enable=false",
-                               "delete.topic.enable=true" })
+                               "auto.create.topics.enable=${topics.autoCreate:false}",
+                               "delete.topic.enable=${topic.delete:true}" })
 public class KafkaStreamsTests {
 
        static final String STREAMING_TOPIC1 = "streamingTopic1";
```
`KafkaStreamsTests` passes.  If I change `${topics.autoCreate:false}` to `${topics.autoCreate:true}`, the test fails as expected (due to the assertion on line 106).  

So it seems to me that the PR works correctly in this case.

Reply:
Exactly! 
That's what I meant.
And that's why I told you that we don't need your hacky test - we are easily covered by the simple modification to the existing test.

Reply:
@artembilan AH I see now!  Ok, great!

Reply:
Sorry, just misunderstanding on my part!